src/sd.jl

```diff
@@ -25,3 +25,113 @@ function addconstraint(m::Model, c::SDVariableConstraint)
     cref = MOI.addconstraint!(m.instance, MOI.VectorOfVariables([instancereference(c.Q[i, j]) for j in 1:n for i in 1:j]), MOI.PositiveSemidefiniteConeTriangle(n))
     return ConstraintRef(m, cref)
 end
+
+##########################################################################
+# SDConstraint is a (dual) semidefinite constraint of the form
+# ∑ cᵢ Xᵢ ⪰ D, where D is a n×n data matrix, cᵢ are scalars,
+# and Xᵢ are n×n symmetric variable matrices. The inequality
+# is taken w.r.t. the psd partial order.
+struct SDConstraint{T<:AbstractJuMPScalar, MT<:AbstractMatrix{T}} <: AbstractConstraint
+    terms::MT
+end
+SDConstraint(terms::MT) where {T<:AbstractJuMPScalar, MT<:AbstractMatrix{T}} = SDConstraint{T, MT}(terms)
+
+"""
+    trimap(i, j)
+
+Returns the vectorized index `k` corresponding to the matrix indices `(i, j)`.
+"""
+function trimap(i::Integer, j::Integer)
+    if i < j
+        trimap(j, i)
+    else
+        div((i-1)*i, 2) + j
+    end
+end
+
+"""
+    invtrimap(i, j)
+
+Returns the matrix indices `(i, j)` (`i <= j`) corresponding to the vectorized index `k`.
+"""
+function invtrimap(k::Integer)
+    i = isqrt(2k)
+    j = k - div((i-1)*i, 2)
+    i, j
+end
+
+struct SDSymInstanceRef{T}
+    sd::MOICON{MOI.VectorAffineFunction{T}, MOI.PositiveSemidefiniteConeTriangle}
+    # symmetry enforcing constraints, it is null if the matrix is symmetric
+    symref::Nullable{MOICON{MOI.VectorAffineFunction{T}, MOI.Zeros}}
+    # if k in symidx then invtrimap(k) has a symmetry enforcing constraint in symref
+    symidx::IntSet
+end
+
+constructconstraint!(x::AbstractMatrix, ::PSDCone) = SDConstraint(x)
+
+_constrain_symmetry(m::Model, c::SDConstraint{T, MT}, sdref) where {T, MT<:Symmetric} = sdref
+function fill_expr!(m::Model, expr::Dict{Variable, T}, terms, coeff=one(T)) where T
```

Comment:
This is general code, shouldn't go into `sd.jl`.

Reply:
Fixed in e5f66d8005708953140b9ede853822d4654825fc